In [1]:
import nltk
from nltk import word_tokenize, pos_tag
from nltk.corpus import stopwords
import requests
import spacy
from bs4 import BeautifulSoup
from urllib.request import Request, urlopen
import re
import random
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.cluster import KMeans

## First compile the reviews

In [2]:
urls = ['https://www.imdb.com/title/tt10954600/reviews?ref_=tt_urv',
        'https://www.imdb.com/title/tt9114286/reviews?ref_=tt_urv',
        'https://www.imdb.com/title/tt10648342/reviews?ref_=tt_urv',
        'https://www.imdb.com/title/tt9419884/reviews?ref_=tt_urv',
        'https://www.imdb.com/title/tt10872600/reviews?ref_=tt_urv',
        'https://www.imdb.com/title/tt9032400/reviews?ref_=tt_urv',
        'https://www.imdb.com/title/tt9376612/reviews?ref_=tt_urv',
        'https://www.imdb.com/title/tt3480822/reviews?ref_=tt_urv',
        'https://www.imdb.com/title/tt6320628/reviews?ref_=tt_urv',
        'https://www.imdb.com/title/tt4154796/reviews?ref_=tt_urv'
       ]
# Antman and the Wasp: Quantumania
# Black Panther: Wakanda Forever
# Thor Love and Thunder
# Doctor Strange in the Multiverse of Madness
# Spider-Man: No Way HOme
# Eternals
# Shang-Chi and the leend of the Ten Rings
# Black Widow
# Spider Man Far From Home
# Avengers Endgame

directory = []
for i in urls:
    req = Request(i)
    html_page = urlopen(req)
    soup = BeautifulSoup(html_page, "lxml")

    all_links = []
    for link in soup.findAll('a'):
        all_links.append(link.get('href'))

    review_links = []
    random.shuffle(all_links)
    for i in all_links:
        if len(review_links) == 10:
            break
        if 'review/' in i:
            review_links.append(i)
    for i in review_links:
        directory.append(i)

In [3]:
print(directory)

['/review/rw8884834/', '/review/rw8875497/', '/review/rw8873784/', '/review/rw8877495/', '/review/rw8879707/', '/review/rw8875320/', '/review/rw8884834/', '/review/rw9002216/', '/review/rw8885390/', '/review/rw8879122/', '/review/rw8844815/', '/review/rw8695632/', '/review/rw8679686/', '/review/rw8683831/', '/review/rw8668856/', '/review/rw8668856/', '/review/rw8670726/', '/review/rw8718626/', '/review/rw8668496/', '/review/rw8664025/', '/review/rw8332897/', '/review/rw8329077/', '/review/rw8315692/', '/review/rw8306334/', '/review/rw8304194/', '/review/rw8331400/', '/review/rw8351608/', '/review/rw8311274/', '/review/rw8524888/', '/review/rw8309971/', '/review/rw8361636/', '/review/rw8117112/', '/review/rw8114393/', '/review/rw8118284/', '/review/rw8143965/', '/review/rw8114820/', '/review/rw8125279/', '/review/rw8114376/', '/review/rw8120384/', '/review/rw8114393/', '/review/rw8009900/', '/review/rw7665817/', '/review/rw7651439/', '/review/rw7648266/', '/review/rw7649796/', '/review/

In [4]:
links = []
for i in directory:
    links.append('https://imdb.com'+i)
    
print(links)

['https://imdb.com/review/rw8884834/', 'https://imdb.com/review/rw8875497/', 'https://imdb.com/review/rw8873784/', 'https://imdb.com/review/rw8877495/', 'https://imdb.com/review/rw8879707/', 'https://imdb.com/review/rw8875320/', 'https://imdb.com/review/rw8884834/', 'https://imdb.com/review/rw9002216/', 'https://imdb.com/review/rw8885390/', 'https://imdb.com/review/rw8879122/', 'https://imdb.com/review/rw8844815/', 'https://imdb.com/review/rw8695632/', 'https://imdb.com/review/rw8679686/', 'https://imdb.com/review/rw8683831/', 'https://imdb.com/review/rw8668856/', 'https://imdb.com/review/rw8668856/', 'https://imdb.com/review/rw8670726/', 'https://imdb.com/review/rw8718626/', 'https://imdb.com/review/rw8668496/', 'https://imdb.com/review/rw8664025/', 'https://imdb.com/review/rw8332897/', 'https://imdb.com/review/rw8329077/', 'https://imdb.com/review/rw8315692/', 'https://imdb.com/review/rw8306334/', 'https://imdb.com/review/rw8304194/', 'https://imdb.com/review/rw8331400/', 'https://im

I compiled a list of reviews for various Marvel movies. 

In [5]:
reviews = []
for i in links:
    url = i
    res = requests.get(url)
    res.encoding = 'utf-8'
    soup = BeautifulSoup(res.text, "lxml")
    reviews.append(soup.select(".text")[0].text)

len(reviews)

100

In [6]:
print(reviews[0])

.........and that literally is a line from the film, summing up how the MCU has become since Covid hit the world. Disney are now jumping the shark 31 films into the universe......Scott Lang and Hope Van Dyne return to continue their adventures as Ant-Man and The Wasp. Together, with Hope's parents Janet and Hank Pym, and Scott's daughter Cassie Lang, the family finds themselves exploring the pretty redundant Quantum Realm, interacting with strange new CGI creatures, and monstrosities, and embarking on an adventure that will push them beyond the limits of boredom, and what they thought possible.I love Ant-Man. The first two films were so so different to anything that Marvel had released, and Paul Rudd was amazing as the titular hero. It has humour that no other Marvel humour had, it was childish, and it was all the better for it. It dared to be different, and the payoff was something fresh, even when the marvel behemoth was becoming a little tired, especially after the debacle that was 

In [7]:
df = pd.DataFrame(reviews, columns=["corpus"])
print(df)

                                               corpus
0   .........and that literally is a line from the...
1   Well, I'll start off by saying that this wasn'...
2   "Ant-Man and the Wasp: Quantumania" is the fir...
3   So, I'm not going to say this is a great Marve...
4   We go down into the quantum world and enter pl...
..                                                ...
95  After Avengers Infinity War, we waited for the...
96  If you're going to watch this movie, avoid any...
97  After watching Infinity war, I was looking for...
98  I had to take several breaks, walk the dogs, p...
99  So it all ends. I have loved most of the MCU, ...

[100 rows x 1 columns]


In [8]:
def preprocess_text(text: str, remove_stopwords: bool) -> str:
    # remove links
    text = re.sub(r"http\S+", "", text)
    # remove special chars and numbers
    text = re.sub("[^A-Za-z]+", " ", text)
    # remove stopwords
    if remove_stopwords:
        # 1. tokenize
        tokens = nltk.word_tokenize(text)
        # 2. check if stopword
        tokens = [w for w in tokens if not w.lower() in stopwords.words("english")]
        # 3. join back together
        text = " ".join(tokens)
    # return text in lower case and stripped of whitespaces
    text = text.lower().strip()
    return text

In [9]:
df['cleaned'] = df['corpus'].apply(lambda x: preprocess_text(x, remove_stopwords=True))

In [10]:
df['cleaned'][0]

'literally line film summing mcu become since covid hit world disney jumping shark films universe scott lang hope van dyne return continue adventures ant man wasp together hope parents janet hank pym scott daughter cassie lang family finds exploring pretty redundant quantum realm interacting strange new cgi creatures monstrosities embarking adventure push beyond limits boredom thought possible love ant man first two films different anything marvel released paul rudd amazing titular hero humour marvel humour childish better dared different payoff something fresh even marvel behemoth becoming little tired especially debacle age ultron wright given cold shoulder could watching something different think blame placed reed shoulders studio final say panache first film boring dull effects meld one like makers taken cantina scene new hope decided make whole movie surrounding adding supposed bigger adversary thaos kang villain majors seems homing campest bone body nothing remotely sinister char

## Load one of the sentiment vocabularies and run the sentiment analyzer. 

In [11]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
sentiment = SentimentIntensityAnalyzer()

In [12]:
df

corpus  \
0   .........and that literally is a line from the...   
1   Well, I'll start off by saying that this wasn'...   
2   "Ant-Man and the Wasp: Quantumania" is the fir...   
3   So, I'm not going to say this is a great Marve...   
4   We go down into the quantum world and enter pl...   
..                                                ...   
95  After Avengers Infinity War, we waited for the...   
96  If you're going to watch this movie, avoid any...   
97  After watching Infinity war, I was looking for...   
98  I had to take several breaks, walk the dogs, p...   
99  So it all ends. I have loved most of the MCU, ...   

                                              cleaned  
0   literally line film summing mcu become since c...  
1   well start saying bad movie great alright like...  
2   ant man wasp quantumania first film phase marv...  
3   going say great marvel film however also say b...  
4   go quantum world enter planet cgi apart actors...  
..                                                ...  
95  avengers infinity war waited avengers endgame ...  
96  going watch movie avoid spoilers even spoiler ...  
97  watching infinity war looking forward much tim...  
98  take several breaks walk dogs play videogames ...  
99  ends loved mcu predictable awaited iw see woul...  

[100 rows x 2 columns]

In [13]:
sentiment.polarity_scores(df['cleaned'][0])

{'neg': 0.118, 'neu': 0.61, 'pos': 0.272, 'compound': 0.9936}

In [14]:
sentiments = []
for i in df['cleaned']:
    sentiments.append(sentiment.polarity_scores(i))

In [15]:
gen_sentiments = []
for i in sentiments:
    if i['compound'] < -0.2:
        gen_sentiments.append('negative')
    elif i['compound'] > 0.2:
        gen_sentiments.append('positive')
    else:
        gen_sentiments.append('neutral')
        

In [16]:
df['gen_sentiments'] = gen_sentiments
print(df['gen_sentiments'])

0     positive
1     positive
2     positive
3     positive
4     positive
        ...   
95    positive
96    positive
97     neutral
98    negative
99    positive
Name: gen_sentiments, Length: 100, dtype: object


Using VADER sentiment analysis, I based the sentiment for each review based on the compound value which is a normalized value based on the negative, neutral, and positive sentiment scores.

## Create Clusters and compute the average, median, high, and low sentiment scores for each cluster. 

In [17]:
# initialize the vectorizer
vectorizer = TfidfVectorizer(sublinear_tf=True, min_df=5, max_df=0.95)
# fit_transform applies TF-IDF to clean texts - we save the array of vectors in X
X = vectorizer.fit_transform(df['cleaned'])

In [18]:
len(vectorizer.get_feature_names())

575

In [19]:
# initialize kmeans with 2 centroids
kmeans = KMeans(n_clusters=2, random_state=42)
# fit the model
kmeans.fit(X)
# store cluster labels in a variable
clusters = kmeans.labels_

In [20]:
clusters

array([1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1], dtype=int32)

In [21]:
def get_top_keywords(n_terms):
    """This function returns the keywords for each centroid of the KMeans"""
    df = pd.DataFrame(X.todense()).groupby(clusters).mean() # groups the TF-IDF vector by cluster
    terms = vectorizer.get_feature_names() # access tf-idf terms
    for i,r in df.iterrows():
        print('\nCluster {}'.format(i))
        print(','.join([terms[t] for t in np.argsort(r)[-n_terms:]])) # for each row of the dataframe, find the n terms that have the highest tf idf score
            
get_top_keywords(10)


Cluster 0
felt,like,really,story,marvel,characters,bad,good,film,movie

Cluster 1
far,time,mcu,marvel,like,one,movie,film,spider,man


In [22]:
cluster0 = []
cluster1 = []
for i in range(0,100):
    if clusters[i] == 0:
        cluster0.append(df['cleaned'][i])
    else:
        cluster1.append(df['cleaned'][i])
len(cluster0)
len(cluster1)

57

In [23]:
cluster0_sentiments = []
for i in cluster0:
    x = sentiment.polarity_scores(i)
    cluster0_sentiments.append(x['compound'])
    
print(cluster0_sentiments)

[0.93, 0.9744, 0.979, -0.9586, 0.9891, 0.0096, -0.836, 0.4345, 0.743, 0.7584, 0.3134, 0.9733, -0.6771, -0.9152, 0.8316, 0.765, -0.3687, -0.9901, -0.9538, 0.765, -0.1779, 0.9952, 0.9487, 0.7906, 0.6249, 0.687, 0.9487, 0.9952, 0.9887, 0.9565, 0.9101, 0.872, 0.872, 0.9869, -0.34, 0.9964, 0.9846, 0.9637, 0.9846, 0.9181, 0.891, 0.9274, 0.5994]


In [24]:
print("The average sentiment for cluster 0 is ", np.average(cluster0_sentiments))
print("The median sentiment for cluster 0 is ", np.median(cluster0_sentiments))
print("The low sentiment for cluster 0 is ", np.min(cluster0_sentiments))
print("The high sentiment for cluster 0 is ", np.max(cluster0_sentiments))

The average sentiment for cluster 0 is  0.5137348837209302
The median sentiment for cluster 0 is  0.872
The low sentiment for cluster 0 is  -0.9901
The high sentiment for cluster 0 is  0.9964


In [25]:
cluster1_sentiments = []
for i in cluster1:
    x = sentiment.polarity_scores(i)
    cluster1_sentiments.append(x['compound'])
    
print(cluster1_sentiments)

[0.9936, 0.9971, 0.9973, 0.9936, 0.9921, 0.9791, -0.0463, 0.4445, 0.9486, 0.9926, 0.9926, 0.5657, 0.9666, 0.9967, 0.9987, 0.7717, 0.9993, 0.9188, -0.591, 0.946, 0.9982, 0.9694, 0.9847, 0.9045, 0.9282, 0.8957, 0.9935, -0.958, 0.7479, 0.765, 0.886, 0.9995, 0.9592, 0.9966, 0.9977, 0.6873, -0.9371, 0.9638, 0.9726, 0.9694, 0.9889, 0.9371, 0.9694, 0.9666, 0.5994, 0.9909, 0.891, 0.9955, 0.5859, 0.8648, 0.9942, -0.505, -0.949, 0.9918, -0.0516, -0.9189, 0.9942]


In [26]:
print("The average sentiment for cluster 1 is ", np.average(cluster1_sentiments))
print("The median sentiment for cluster 1 is ", np.median(cluster1_sentiments))
print("The low sentiment for cluster 1 is ", np.min(cluster1_sentiments))
print("The high sentiment for cluster 1 is ", np.max(cluster1_sentiments))

The average sentiment for cluster 1 is  0.7004666666666667
The median sentiment for cluster 1 is  0.9666
The low sentiment for cluster 1 is  -0.958
The high sentiment for cluster 1 is  0.9995


There does not seem to be much difference in the sentiments between the clusters. The averages and medians are close, and both clusters have very low sentiment reviews and very high sentiment reviews. Maybe we will see more separation with more clusters.

In [27]:
# initialize kmeans with 2 centroids
kmeans = KMeans(n_clusters=9, random_state=42)
# fit the model
kmeans.fit(X)
# store cluster labels in a variable
clusters = kmeans.labels_

In [28]:
def get_top_keywords(n_terms):
    """This function returns the keywords for each centroid of the KMeans"""
    df = pd.DataFrame(X.todense()).groupby(clusters).mean() # groups the TF-IDF vector by cluster
    terms = vectorizer.get_feature_names() # access tf-idf terms
    for i,r in df.iterrows():
        print('\nCluster {}'.format(i))
        print(','.join([terms[t] for t in np.argsort(r)[-n_terms:]])) # for each row of the dataframe, find the n terms that have the highest tf idf score
            
get_top_keywords(10)


Cluster 0
years,old,movie,mcu,life,marvel,man,shang,chi,awkwafina

Cluster 1
great,mcu,last,time,humour,man,far,see,film,spider

Cluster 2
scarlett,film,much,story,johansson,pugh,florence,movie,black,widow

Cluster 3
except,say,time,one,much,going,movie,bad,watching,anything

Cluster 4
movie,forever,film,boseman,chadwick,namor,story,black,wakanda,panther

Cluster 5
mcu,people,good,one,film,films,ant,movie,man,like

Cluster 6
fans,really,written,enjoy,excellent,pace,longer,love,fight,sequences

Cluster 7
know,movie,man,avengers,holland,spider,tom,far,mysterio,peter

Cluster 8
multiverse,sequel,shows,pacing,hand,version,expect,endless,quickly,spoil


In [29]:
clusters

array([5, 5, 5, 5, 5, 5, 5, 3, 5, 5, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 0, 5,
       3, 3, 5, 6, 5, 5, 2, 1, 0, 5, 8, 3, 5, 3, 5, 5, 1, 8, 5, 7, 7, 7,
       4, 0, 1, 5, 0, 0, 5, 0, 6, 5, 7, 0, 0, 3, 6, 5, 0, 4, 5, 2, 2, 0,
       4, 0, 0, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 7, 7, 1, 7, 5, 1, 7, 7,
       7, 7, 1, 5, 1, 0, 5, 5, 3, 3, 3, 1], dtype=int32)

In [30]:
cluster0 = []
cluster1 = []
cluster2 = []
cluster3 = []
cluster4 = []
cluster5 = []
cluster6 = []
cluster7 = []
cluster8 = []

for i in range(0,100):
    if clusters[i] == 0:
        cluster0.append(df['cleaned'][i])
    if clusters[i] == 1:
        cluster1.append(df['cleaned'][i])
    if clusters[i] == 2:
        cluster2.append(df['cleaned'][i])
    if clusters[i] == 3:
        cluster3.append(df['cleaned'][i])
    if clusters[i] == 4:
        cluster4.append(df['cleaned'][i])
    if clusters[i] == 5:
        cluster5.append(df['cleaned'][i])
    if clusters[i] == 6:
        cluster6.append(df['cleaned'][i])
    if clusters[i] == 7:
        cluster7.append(df['cleaned'][i])
    if clusters[i] == 8:
        cluster8.append(df['cleaned'][i])
print(len(cluster0))
print(len(cluster1))
print(len(cluster2))
print(len(cluster3))
print(len(cluster4))
print(len(cluster5))
print(len(cluster6))
print(len(cluster7))
print(len(cluster8))

14
8
13
9
13
27
3
11
2


In [31]:
cluster0_sentiments = []
for i in cluster0:
    x = sentiment.polarity_scores(i)
    cluster0_sentiments.append(x['compound'])
        
print(cluster0_sentiments)

cluster1_sentiments = []
for i in cluster1:
    x = sentiment.polarity_scores(i)
    cluster1_sentiments.append(x['compound'])
        
print(cluster1_sentiments)

cluster2_sentiments = []
for i in cluster2:
    x = sentiment.polarity_scores(i)
    cluster2_sentiments.append(x['compound'])
        
print(cluster2_sentiments)

cluster3_sentiments = []
for i in cluster3:
    x = sentiment.polarity_scores(i)
    cluster3_sentiments.append(x['compound'])
        
print(cluster3_sentiments)

cluster4_sentiments = []
for i in cluster4:
    x = sentiment.polarity_scores(i)
    cluster4_sentiments.append(x['compound'])
        
print(cluster4_sentiments)


cluster5_sentiments = []
for i in cluster5:
    x = sentiment.polarity_scores(i)
    cluster5_sentiments.append(x['compound'])
        
print(cluster5_sentiments)

cluster6_sentiments = []
for i in cluster6:
    x = sentiment.polarity_scores(i)
    cluster6_sentiments.append(x['compound'])
        
print(cluster6_sentiments)

cluster7_sentiments = []
for i in cluster7:
    x = sentiment.polarity_scores(i)
    cluster7_sentiments.append(x['compound'])
        
print(cluster7_sentiments)

cluster8_sentiments = []
for i in cluster8:
    x = sentiment.polarity_scores(i)
    cluster8_sentiments.append(x['compound'])
        
print(cluster8_sentiments)

[0.0096, 0.7717, 0.9282, -0.958, 0.7479, 0.765, 0.9995, 0.6249, 0.9887, 0.9592, 0.9966, 0.9977, -0.34, -0.505]
[0.8316, 0.946, 0.8957, 0.9371, 0.5994, 0.9274, 0.9942, 0.9942]
[-0.9152, 0.872, 0.872, 0.9964, 0.6873, 0.9846, -0.9371, 0.9638, 0.9637, 0.9726, 0.9846, 0.9181, 0.891]
[-0.9586, 0.4345, 0.743, 0.9188, -0.3687, 0.687, 0.5994, -0.0516, -0.9189]
[-0.0463, 0.4445, 0.9486, 0.9891, 0.9926, 0.9926, 0.5657, 0.9666, 0.9967, 0.9987, 0.9045, 0.9565, 0.9869]
[0.9936, 0.93, 0.9971, 0.9973, 0.9744, 0.979, 0.9936, 0.9921, 0.9791, -0.836, 0.7584, 0.9733, -0.6771, 0.9993, -0.591, -0.9901, -0.9538, -0.1779, 0.9935, 0.9952, 0.7906, 0.9952, 0.9101, 0.9666, 0.8648, -0.949, 0.9918]
[0.3134, 0.9487, 0.9487]
[0.9982, 0.9694, 0.9847, 0.886, 0.9694, 0.9889, 0.9694, 0.9909, 0.891, 0.9955, 0.5859]
[0.765, 0.765]


In [32]:
sentiments_list = [cluster0_sentiments,cluster1_sentiments,cluster2_sentiments,
                   cluster3_sentiments,cluster4_sentiments,cluster5_sentiments,
                   cluster6_sentiments,cluster7_sentiments,cluster8_sentiments]

In [33]:
averages = []
medians = []
lows = []
highs = []

for i in sentiments_list:
    averages.append(np.average(i))
    medians.append(np.median(i))
    lows.append(np.min(i))
    highs.append(np.max(i))
    
print("The averages are ", averages)
print("The medians are ", medians)
print("The lows are ", lows)
print("The highs are ", highs)

The averages are  [0.49900000000000005, 0.8907, 0.6349076923076923, 0.12054444444444444, 0.8228230769230771, 0.5148185185185186, 0.7369333333333333, 0.9299363636363637, 0.765]
The medians are  [0.7683500000000001, 0.93225, 0.9181, 0.4345, 0.9666, 0.9733, 0.9487, 0.9694, 0.765]
The lows are  [-0.958, 0.5994, -0.9371, -0.9586, -0.0463, -0.9901, 0.3134, 0.5859, 0.765]
The highs are  [0.9995, 0.9942, 0.9964, 0.9188, 0.9987, 0.9993, 0.9487, 0.9982, 0.765]


We see that all of the clusters have positive sentiment on average, though some clusters are significantly lower than others. Clusters 0, 2, 7, and 8 have low average sentiment values. All clusters have very positive sentiment reviews and low sentiment reviews except clusters 1 and 5 which have a low sentiment of 0 and 0.5994, respectively, which are pretty high for minimums. 